src/puppetlabs/ring_middleware/core.clj

```diff
@@ -30,22 +164,138 @@
+                      :plain (plain-response code (:message e))))]
+     (fn [request]
+       (sling/try+ (handler request)
+             (catch
+               #(contains? #{:request-data-invalid
+                             :user-data-invalid
+                             :data-invalid
+                             :service-status-version-not-found}
+                           (:type %))
+               e
+               (response e)))))))
+
+(schema/defn ^:always-validate wrap-service-unavailable :- IFn
+  "A ring middleware that catches slingshot errors thrown by
+  throw-service-unavailabe!, logs the error and returns a 503 ring response."
+  ([handler :- IFn]
+   (wrap-service-unavailable handler :json))
+  ([handler :- IFn
+    type :- ResponseType]
+   (let [code 503
+         response (fn [e]
+                    (log/error "Service Unavailable:" (:message e))
+                    (case type
+                      :json (json-response code {:error e})
+                      :plain (plain-response code (:message e))))]
+     (fn [request]
+       (sling/try+ (handler request)
+         (catch service-unavailable? e
+                (response e)))))))
+
+(schema/defn ^:always-validate wrap-bad-request :- IFn
+  "A ring middleware that catches slingshot errors thrown by
+  throw-bad-request!, logs the error and returns a 503 ring response."
+  ([handler :- IFn]
+   (wrap-bad-request handler :json))
+  ([handler :- IFn
+    type :- ResponseType]
+   (let [code 400
+         response (fn [e]
+                    (log/error "Bad Request:" (:message e))
+                    (case type
+                      :json (json-response code {:error e})
+                      :plain (plain-response code (:message e))))]
+     (fn [request]
+       (sling/try+ (handler request)
+         (catch bad-request? e
+                (response e)))))))
+
+(schema/defn ^:always-validate wrap-schema-errors :- IFn
+  "A ring middleware that catches schema errors and returns a 500
+  response with the details"
+  ([handler :- IFn]
+   (wrap-schema-errors handler :json))
+  ([handler :- IFn
+    type :- ResponseType]
+   (let [code 500
+         response (fn [e]
+                    (let [msg (str "Something unexpected happened:"
+                                   (select-keys (.getData e) [:error :value :type]))]
+                      (log/error msg)
+                      (case type
+                        :json (json-response code
+                                             {:error
+                                              {:type :application-error
+                                               :message msg}})
+                        :plain (plain-response code msg))))]
+     (fn [request]
+       ;; We cannot use slingshot.slingshot/try+ (as of 0.12.2) because
+       ;; it will unwrap an IExceptionInfo into a map to match against,
+       ;; while schema will raise a regular ExceptionInfo class to match.
```

Comment:
Well, damn. I definitely didn't understand something about what I was doing. Thanks for both the correction and the example repl snippet. I'll try and figure out what I wasn't understanding and update this.